timesketch/ui/graphs/cytoscape.component.ts

```diff
@@ -0,0 +1,224 @@
+import {
+  Component, ElementRef, Input, Output, OnChanges, SimpleChanges, EventEmitter,
+  AfterViewInit,
+} from '@angular/core'
+import * as cytoscape from 'cytoscape'
+
+type Opt = Cy.CytoscapeOptions
+
+interface CytoscapeEvent<T> {
+  cy: Cy.Core
+  target: T
+  type: string
+  namespace: string
+  timeStamp: number
+}
+
+interface CytoscapeLayoutEvent extends CytoscapeEvent<Cy.Core> {
+  layout: Cy.Layouts
+}
+
+interface CytoscapeInputEvent<E, T> extends CytoscapeEvent<T> {
+    position: Cy.Position
+    renderedPosition: Cy.Position
+    originalEvent: E
+}
+
+type Evt = {
+  [K in keyof DocumentEventMap]: CytoscapeInputEvent<DocumentEventMap[K], Cy.Core | Cy.SingularElement>
+} & {
+  taphold: CytoscapeInputEvent<TouchEvent, Cy.Core | Cy.SingularElement>
+  boxstart: CytoscapeInputEvent<Event, Cy.Core>
+  boxend: CytoscapeInputEvent<Event, Cy.Core>
+  boxselect: CytoscapeInputEvent<Event, Cy.SingularElement>
+  box: CytoscapeInputEvent<Event, Cy.SingularElement>
+  layoutstart: CytoscapeLayoutEvent
+  layoutstop: CytoscapeLayoutEvent
+  layoutready: CytoscapeLayoutEvent
+  ready: CytoscapeEvent<Cy.Core>
+  destroy: CytoscapeEvent<Cy.Core>
+  render: CytoscapeEvent<Cy.Core>
+  resize: CytoscapeEvent<Cy.Core>
+}
+
+const immutable_options = [
+  'headless', 'styleEnabled', 'hideEdgesOnViewport', 'hideLabelsOnViewport',
+  'textureOnViewport', 'motionBlur', 'touchTapThreshold',
+  'desktopTapThreshold', 'motionBlurOpacity', 'wheelSensitivity',
+  'pixelRatio', 'selectionType',
+]
+
+const mutable_options = [
+  'elements', 'style', 'layout', 'zoom', 'pan', 'minZoom', 'maxZoom',
+  'zoomingEnabled', 'userZoomingEnabled', 'panningEnabled',
+  'userPanningEnabled', 'boxSelectionEnabled', 'autolock', 'autoungrabify',
+  'autounselectify',
+]
+
+/**
+ * Component that embeds Cytoscape. Input properties correspond to Cytoscape
+ * initialization options. They can be changed at any time and the changes will
+ * be reflected. If an immutable option is changed, Cytoscape instance will be
+ * rebuilt. Properties {@link CytoscapeComponent#zoom} and {@link CytoscapeComponent#pan}
+ * can be two-way bound using the banana-in-box syntax.
+ * @see {@link http://js.cytoscape.org/#core/initialisation}
+ */
+@Component({
+  selector: 'ts-graphs-cytoscape',
+  template: '<div></div>',
+})
+export class CytoscapeComponent implements OnChanges, AfterViewInit {
+  private element: HTMLElement
+  private _cy: Cy.Core
+
+  get cy(): Cy.Core {
```

Comment:
Please add a comment here. I assume you expose cy via a property to make it read-only?

Reply:
Yes.

Reply:
Done.